workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/utilities/TokenBuilder.kt

```diff
@@ -0,0 +1,157 @@
+     *
+     * @param doubleAmount The [Double] we will use to set the [doubleAmount] member property.
+     */
+    fun withAmountValue(doubleAmount: Double) = initializeAmountOrThrow { this.doubleAmount = doubleAmount }
+
+    /**
+     * Initialize the [bigDecimalAmount] member property of the builder. Throws an exception if another
+     * possible amount has already been initialized.
+     *
+     * @param bigDecimalAmount The [BigDecimal] we will use to set the [bigDecimalAmount] member property.
+     */
+    fun withAmountValue(bigDecimalAmount: BigDecimal) = initializeAmountOrThrow { this.bigDecimalAmount = bigDecimalAmount }
+
+    /**
+     * Replicates the Kotlin DSL [of] infix function. Supplies a [TokenType] to the builder
+     * which will be used to build an [Amount] of a [TokenType].
+     *
+     * @param t The token type that will be used to build an [Amount] of a [TokenType]
+     */
+    fun <T: TokenType> of(t: T): TokenBuilder = this.apply {
+        this.tokenType = t
+    }
+
+    /**
+     * Replicates the Kotlin DSL [issuedBy] infix function. Supplies a [Party] to the builder
+     * representing the identity of the issuer of an [Amount] of an [IssuedTokenType].
+     *
+     * @param party The issuing identity that will be used to build an [Amount] of an [IssuedTokenType]
+     */
+    fun issuedBy(party: Party): TokenBuilder = this.apply {
+        this.issuer = party
+    }
+
+    /**
+     * Replicates the Kotlin DSL [heldBy] infix function. Supplies a [Party] to the builder
+     * representing the identity of the holder of a new fungible token.
+     *
+     * @param party The identity of the holder that will be used to build an [Amount] of an [IssuedTokenType].
+     */
+    fun heldBy(party: Party) = this.apply {
+        this.holder = party
+    }
+
+    /**
+     * Builds an [Amount] of a [TokenType]. This function will throw a [TokenBuilderException]
+     * exception if the appropriate builder methods have not been called: [withAmountValue], [of].
+     */
+    fun buildAmountTokenType(): Amount<TokenType> = try {
+        when (getInitializedAmount()) {
+            is Long -> { amount(this.longAmount!!, tokenType) }
+            is Int -> amount(this.intAmount!!, tokenType)
+            is Double -> { amount(this.doubleAmount!!, tokenType) }
+            is BigDecimal -> { amount(this.bigDecimalAmount!!, tokenType) }
+            else -> { throw TokenBuilderException("The builder has failed, an appropriate amount type was not provided.") }
+        }
+    } catch (ex: UninitializedPropertyAccessException) {
+        throw TokenBuilderException("An amount value has not been provided to the builder.")
+    }
+
+    /**
+     * Builds an [Amount] of an [IssuedTokenType]. This function will throw a [TokenBuilderException]
+     * if the appropriate builder methods have not been called: [withAmountValue], [of], [issuedBy].
+     */
+    fun buildAmountIssuedTokenType(): Amount<IssuedTokenType> = when {
```

Comment:
Any reason why not to call it just `buildIssuedToken`?

Reply:
I think `buildIssuedToken` could be confusing because what's returned isn't actually a token yet. I agree that `buildAmountIssuedTokenType` is a little verbose, but I think in this case it's appropriate. Happy to change if you think!

Reply:
What about `buildFungibleToken` then? How different is it from others?

Reply:
This is the only method that returns a state - specifically a `FungibleToken` which extends `FungibleState<IssuedTokenType>` while buildAmountIssuedTokenType and buildAmountTokenType both just return `Amounts`